.github/workflows/deploy.yml

```diff
@@ -0,0 +1,155 @@
+name: Deploy
+on:
+  push:
+    branches: [main, release]
+    paths-ignore: ['**.md']
+  pull_request:
+    branches: [main]
+    paths-ignore: ['**.md']
+
+defaults:
+  run:
+    shell: bash
+
+permissions:
+  pull-requests: write
+
+jobs:
+  deploy:
+    name: Deploy
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v5
+      - uses: ./.github/workflows/setup
+      - name: Install wrangler globally
+        run: |
+          # the wrangler action expects a global installation, so we better reuse what we've in our project
+          wrangler_version=$(cat package.json | jq -r '.pnpm.overrides.wrangler')
+          pnpm install -g "wrangler@${wrangler_version}"
+
+      - uses: dorny/paths-filter@v3
+        id: changes
+        with:
+          filters: |
+            api: &api
+              - 'api/**'
+              - 'shared/**'
+            cdn: &cdn
+              - *api
+              - 'components/**'
+              - 'embed/**'
+              - 'cdn/**'
+            frontend:
+              - *api
+              - *cdn
+              - 'frontend/**'
+
+      - name: Get deploy mode
+        id: deploy-mode
+        run: |
+          mode="preview"
+          if [[ "$EVENT_NAME" == 'push' ]]; then
+            [[ "$GH_REF" == 'refs/heads/main' ]] && mode="staging"
+            [[ "$GH_REF" == 'refs/heads/release' ]] && mode="production"
+          fi
+          echo "mode=${mode}" >> $GITHUB_OUTPUT
+        env:
+          GH_EVENT_NAME: ${{ github.event_name }}
+          GH_REF: ${{ github.ref }}
+
+      #region Deploy
+
+      # cdn (embed) needs: API_URL, APP_URL (infer from API_URL to prevent cycle)
+      # frontend (app) needs: CDN_URL, ?API_URL
+      # api needs: nothing (APP_URL is passed from embed)
+      #
+      # build/deploy sequence:
+      # 1. api
+      # 2. cdn (embed)
+      # 3. frontend
+
+      - name: Deploy API
+        uses: ./.github/workflows/deploy-worker
+        id: api
+        with:
+          name: 'api'
+          changed: ${{ steps.changes.outputs.api }}
+          mode: ${{ steps.deploy-mode.outputs.mode }}
+          apiToken: ${{ secrets.CLOUDFLARE_API_TOKEN }}
+          accountId: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
+          stagingUrl: 'https://staging-publisher-tools-api.webmonetization.workers.dev'
+          productionUrl: 'https://api.webmonetization.org'
```

Comment:
Probably worth having the URLs as repository variables? This way if something changes you will only have to update the repository variable.